### DISCUSSION TITLE
使用 C# 撰寫 ModbusTCP client | 全端開發人員天梯

### DISCUSSION THREAD

**Discussion by giscus** [wellwind/wellwind.github.io#84] [Announcements]
# 使用 C# 撰寫 ModbusTCP client | 全端開發人員天梯

在之前的文章「簡介 Modbus TCP」我們簡單介紹了工業上常見且簡單的通訊方式，Modbus 以及 ModbusTCP，由於 ModbusTCP 本身非常簡單，因此就算不靠任何 library，只要懂得撰寫簡單的 TCP 程式就可以自己實作一個 ModbusTCP client，所以今天就以 C# 為例，實作一個簡單具有讀取功能的 ModbusTCP client。

https://fullstackladder.dev/blog/2022/11/12/modbustcp-client-using-csharp/

**Comment by huuuang** [comment]
// See https://aka.ms/new-console-template for more information

using System.Net.Sockets;
using System.Text;

Console.WriteLine("Program statr!");

try
{
    //建立tcpClient, 并连接
    var tcpClient = new TcpClient();
    await tcpClient.ConnectAsync("127.0.0.1", 502); /*原文多了一个下划线*/

    //发送请求封包
    Int16 address = 2000;                           /*需要指定为16位, 我这里默认是32位, 否则reverse后是16#0000 07d0, 返回0x83*/
    Int16 quantity = 5;
    var startingAddress = BitConverter.GetBytes(address);
    var quantityOfRegisters = BitConverter.GetBytes(quantity);

    if (BitConverter.IsLittleEndian)
    {
        startingAddress = startingAddress.Reverse().ToArray();
        quantityOfRegisters = quantityOfRegisters.Reverse().ToArray();
    }
    var request = new byte[]
        {
        0x00,0x01,                                       //Transaction Identifiter, 连续数字或随机数字都可以
        0x00,0x00,                                       //Protocol Identifier, 固定0
        0x00,0x06,                                       //Length, 接下来有6byte的资料
        0x01,                                            //Unit Identifier
        0x03,                                            //Function Code
        startingAddress[0],startingAddress[1],           //Starting address (2002转成2 bytes的结果)
        quantityOfRegisters[0],quantityOfRegisters[1]    //Quantity of Registers (1转成2bytes的结果)
        };
    Console.WriteLine("发送完毕.");
    Console.WriteLine(BitConverter.ToString(request));

    //获取网络流
    var stream = tcpClient.GetStream(); /*原文没实例化*/
    //发送数据
    await stream.WriteAsync(request, 0, request.Length);

    var response = new byte[1024];
    var readCount = await stream.ReadAsync(response, 0, response.Length);
    // 比較完整的寫法是把 Transaction Identifier、Function Code 等都抓出來檢查
    // 尤其是 Function Code，因為當有 error 時就不會是 0x03，而是 0x83 了
    // 不過這邊就先省略，以抓到我們的目標資料為主，那麼就是最後的 quantity * 2 個 bytes
    var result = response.Take(readCount).TakeLast(quantity * 2).ToArray();

    Console.WriteLine("接收完毕");
    Console.WriteLine(BitConverter.ToString(response));
    Console.WriteLine("读取数量:" + readCount);
}
catch (Exception ex)
{
    Console.WriteLine("发生错误:" + ex.Message);
}


2023.07.19
